tests/Unit/ControlSystem/Test_PiecewisePolynomial.cpp

```diff
@@ -1,101 +1,79 @@
-  double t = 0.0;
-  const double dt = 0.6;
-  const double final_time = 4.0;
-  constexpr size_t deriv_order = 3;
-
-  // test two component system (x**3 and x**2)
-  const std::array<DataVector, deriv_order + 1> init_func{
-      {{0.0, 0.0}, {0.0, 0.0}, {0.0, 2.0}, {6.0, 0.0}}};
-  FunctionsOfTime::PiecewisePolynomial<deriv_order> f_of_t_derived(t,
-                                                                   init_func);
-  FunctionOfTime& f_of_t = f_of_t_derived;
-
+namespace {
+template <size_t DerivOrder>
+void test(const gsl::not_null<FunctionOfTime*> f_of_t,
+          const gsl::not_null<FunctionsOfTime::PiecewisePolynomial<DerivOrder>*>
+              f_of_t_derived,
+          double t, const double dt, const double final_time) noexcept {
   while (t < final_time) {
-    const auto lambdas0 = f_of_t.func_and_2_derivs(t);
+    const auto lambdas0 = f_of_t->func_and_2_derivs(t);
     CHECK(approx(lambdas0[0][0]) == cube(t));
     CHECK(approx(lambdas0[0][1]) == square(t));
     CHECK(approx(lambdas0[1][0]) == 3.0 * square(t));
     CHECK(approx(lambdas0[1][1]) == 2.0 * t);
     CHECK(approx(lambdas0[2][0]) == 6.0 * t);
     CHECK(approx(lambdas0[2][1]) == 2.0);
 
-    const auto lambdas1 = f_of_t.func_and_deriv(t);
+    const auto lambdas1 = f_of_t->func_and_deriv(t);
     CHECK(approx(lambdas1[0][0]) == cube(t));
     CHECK(approx(lambdas1[0][1]) == square(t));
     CHECK(approx(lambdas1[1][0]) == 3.0 * square(t));
     CHECK(approx(lambdas1[1][1]) == 2.0 * t);
 
-    const auto lambdas2 = f_of_t.func(t);
+    const auto lambdas2 = f_of_t->func(t);
     CHECK(approx(lambdas2[0][0]) == cube(t));
     CHECK(approx(lambdas2[0][1]) == square(t));
 
     t += dt;
-    f_of_t_derived.update(t, {6.0, 0.0});
+    f_of_t_derived->update(t, {6.0, 0.0});
   }
   // test time_bounds function
-  const auto t_bounds = f_of_t.time_bounds();
+  const auto t_bounds = f_of_t->time_bounds();
   CHECK(t_bounds[0] == 0.0);
   CHECK(t_bounds[1] == 4.2);
 }
 
-SPECTRE_TEST_CASE(
-    "Unit.ControlSystem.FunctionsOfTime.PiecewisePolynomial.NonConstDeriv",
-    "[ControlSystem][Unit]") {
-  double t = 0.0;
-  const double dt = 0.6;
-  const double final_time = 4.0;
-  constexpr size_t deriv_order = 2;
-
-  // initally x**2, but update with non-constant 2nd deriv
-  const std::array<DataVector, deriv_order + 1> init_func{
-      {{0.0}, {0.0}, {2.0}}};
-  FunctionsOfTime::PiecewisePolynomial<deriv_order> f_of_t_derived(t,
-                                                                   init_func);
-  FunctionOfTime& f_of_t = f_of_t_derived;
-
+template <size_t DerivOrder>
+void test_non_const_deriv(
+    const gsl::not_null<FunctionOfTime*> f_of_t,
+    const gsl::not_null<FunctionsOfTime::PiecewisePolynomial<DerivOrder>*>
+        f_of_t_derived,
+    double t, const double dt, const double final_time) noexcept {
   while (t < final_time) {
     t += dt;
-    f_of_t_derived.update(t, {3.0 + t});
+    f_of_t_derived->update(t, {3.0 + t});
```

Comment:
I know this is neither your code or new code, so this shouldn't hold up the PR... but I don't understand what is happening here. Why is there a separate pointer to the `f_of_t`, of the derived type, for calling `update`?

Reply:
I don't remember the conversation exactly, but my guess is that it tests that the updating via the base class works correctly.